File: java-algorithm/src/main/java/com/example/algorithm/sort/HeapSort.java
package com.example.algorithm.sort;

import static com.example.algorithm.util.ElementUtil.swap;

public class HeapSort {
    public static int[] heapSort(int[] array) {
        int heapSize = array.length;
        buildMaxHeap(array);
        for (int i = heapSize - 1; i > 0; i--) {
            swap(array, 0, i);
            heapSize--;
            maxHeapify(array, heapSize, 0);
        }
        return array;
    }

    private static void buildMaxHeap(int[] array) {
        int heapSize = array.length;
        int loopIndex = Math.subtractExact(Math.floorDiv(heapSize, 2), 1);

        for (int i = loopIndex; i > -1; i--) {
            maxHeapify(array, heapSize, i);
        }
    }

    private static void maxHeapify(int[] array, int heapSize, int index) {
        int largest = index;
        int leftIndex = 2 * index + 1; // 2n + 1
        int rightIndex = 2 * index + 2; // 2n + 2

        if ((leftIndex < heapSize) && (array[largest] < array[leftIndex])) {
            largest = leftIndex;
        }
        if ((rightIndex < heapSize) && (array[largest] < array[rightIndex])) {
            largest = rightIndex;
        }
        if (largest != index) {
            swap(array, index, largest);
            maxHeapify(array, heapSize, largest);
        }
    }
}

File: go-algorithm/pkg/sort/heap_sort.go
package sort

import (
	"golang.org/x/exp/constraints"
)

func HeapSort[T constraints.Integer](arr []T) {
	n := len(arr)
	buildMaxHeap(arr)
	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		maxHeapify(arr, 0, i)
	}
}

func buildMaxHeap[T constraints.Integer](arr []T) {
	n := len(arr)
	for i := n/2 - 1; i >= 0; i-- {
		maxHeapify(arr, i, n)
	}
}

func maxHeapify[T constraints.Integer](arr []T, i, n int) {
	largest := i
	for {
		left := 2*i + 1
		right := 2*i + 2

		if left < n && arr[largest] < arr[left] {
			largest = left
		}
		if right < n && arr[largest] < arr[right] {
			largest = right
		}
		if largest == i {
			break
		}

		arr[i], arr[largest] = arr[largest], arr[i]
		i = largest
	}
}
